libavcodec/x86/vvc/vvc_sad.asm

```diff
@@ -0,0 +1,265 @@
+        movd              eax, xm0
+
+    RET
+
+cglobal vvc_sad_32_16bpc, 6, 9, 13, src1, src2, dx, dy, block_w, block_h, off1, off2, row_idx
+    INIT_OFFSET src1q, src2q, dxq, dyq, off1q, off2q
+    pxor                 m3, m3
+
+.loop_height:
+
+    mov               off1q, src1q
+    mov               off2q, src2q
+
+    .loop_row:
+        movu              xm0, [src1q]
+        movu              xm1, [src2q]
+        MIN_MAX_SAD       xm2, xm0, xm1
+        vpmovzxwd          m1, xm1
+        vpaddd             m3, m1
+
+        movu              xm5, [src1q + 16]
+        movu              xm6, [src2q + 16]
+        MIN_MAX_SAD       xm7, xm5, xm6
+        vpmovzxwd          m6, xm6
+        vpaddd             m3, m6
+
+        movu              xm8, [src1q + 32]
+        movu              xm9, [src2q + 32]
+        MIN_MAX_SAD       xm10, xm8, xm9
+        vpmovzxwd          m9, xm9
+        vpaddd             m3, m9
+
+        movu              xm11, [src1q + 48]
+        movu              xm12, [src2q + 48]
+        MIN_MAX_SAD       xm13, xm11, xm12
+        vpmovzxwd          m12, xm12
+        vpaddd             m3, m12
+
+    lea             src1q, [off1q + 2 * 128 * 2] 
+    lea             src2q, [off2q + 2 * 128 * 2]
+
+    sub block_hq, 2
+    jg .loop_height
+
+    HORIZONTAL_ADD xm0, xm3, m3
+    movd              eax, xm0
+
+    RET
+
+cglobal vvc_sad_64_16bpc, 6, 9, 13, src1, src2, dx, dy, block_w, block_h, off1, off2, row_idx
+    INIT_OFFSET src1q, src2q, dxq, dyq, off1q, off2q
+    pxor                 m3, m3
+
+.loop_height:
+
+    mov               off1q, src1q
+    mov               off2q, src2q
+    mov            row_idxd, 2
+
+    .loop_row:
+        movu              xm0, [src1q]
+        movu              xm1, [src2q]
+        MIN_MAX_SAD       xm2, xm0, xm1
+        vpmovzxwd          m1, xm1
+        vpaddd             m3, m1
+
+        movu              xm5, [src1q + 16]
+        movu              xm6, [src2q + 16]
+        MIN_MAX_SAD       xm7, xm5, xm6
+        vpmovzxwd          m6, xm6
+        vpaddd             m3, m6
+
+        movu              xm8, [src1q + 32]
+        movu              xm9, [src2q + 32]
+        MIN_MAX_SAD      xm10, xm8, xm9
+        vpmovzxwd          m9, xm9
+        vpaddd             m3, m9
+
+        movu              xm11, [src1q + 48]
+        movu              xm12, [src2q + 48]
+        MIN_MAX_SAD       xm13, xm11, xm12
+        vpmovzxwd          m12, xm12
+        vpaddd              m3, m12
+
+        add              src1q, 64
+        add              src2q, 64
+        dec           row_idxd
+        jg  .loop_row
+
+    lea             src1q, [off1q + 2 * 128 * 2] 
+    lea             src2q, [off2q + 2 * 128 * 2]
+
+    sub block_hq, 2
+    jg .loop_height
+
+    HORIZONTAL_ADD xm0, xm3, m3
+    movd              eax, xm0
+
+    RET
+
+cglobal vvc_sad_128_16bpc, 6, 9, 13, src1, src2, dx, dy, block_w, block_h, off1, off2, row_idx
```

Comment:
looks like there are a lot of the same operations of 32/64/128? Can we merge them into one macro?

Reply:
Created the macro here 
https://github.com/stone-d-chen/ffvvc/blob/0c68dad30567785304760ef6aeb2a3e439a4d6a2/libavcodec/x86/vvc/vvc_sad.asm#L72